### ISSUE TITLE
HttpRouter takes incorrect route

### ISSUE THREAD

**Issue by pcoltau** [glock45/swifter#355]
When having overlapping paths, the HttpRouter will pick the incorrect path.

I have the following test, which fails:

```
let handler: ((HttpRequest) -> HttpResponse) = { _ in
    return HttpResponse.accepted
}

let router = HttpRouter()
router.register("GET", path: "a/b", handler: handler)
router.register("GET", path: "a/:id/c", handler: handler)

let foundHandler = router.route("GET", path: "a/b")
XCTAssertNotNil(foundHandler)

let foundSecondHandler = router.route("GET", path: "a/b/c")
XCTAssertNotNil(foundSecondHandler)
```

The problem is in the implementation of `HttpRouter.findHandler()`, which has two issues:

1. Path parameters are chosen before a path segment, which means, in the first assertion in the above test, the route `"a/:id/c"` will be chosen over `"a/b"`. This can be fixed, in `HttpRouter.findHandler()`, by looking for the `pathToken` before checking for `variableNodes`.

However, with the first issue fixed, the `HttpRouter` will still not correctly find the handler for the second assertion in the above test. The reason for this is the second issue:

2. The `HttpRouter.findHandler()` does not reevaluate if a `nil` handler is found. If the found handler is `nil`, it might indicate that the routing was incorrect, and a different route should be selected. One could introduce `if let` optional unwrapping of the handler returned by calls to `findHandler()` in order to not return the found handler, but instead just continue the search for a fitting route. However, the generator used in the `findHandler()` function is "global" and has already been advanced to the next `pathToken`.

**Comment by Vkt0r** [comment]
Hey, @pcoltau thanks for bringing this up and for creating the unit test for it! Yes, it's a bug in the traversal of the tree for the path. I'll try to fix it as soon as I can.

**Comment by pcoltau** [comment]
Thank you for fixing this issue! 

When will it be included in a release?

**Comment by Vkt0r** [comment]
I'll try to make a release this week.

**Comment by Vkt0r** [comment]
@pcoltau A new release [`1.4.6`](https://github.com/httpswift/swifter/releases/tag/1.4.6) is up.

**Comment by pcoltau** [comment]
I've tested the new version, and unfortunately, there is still an issue.

If I change the route registation in my test to the following, then the test will fail:

```
router.register("GET", path: "a/:id", handler: handler)
router.register("GET", path: "a/:id/c", handler: handler)
```

**Comment by Vkt0r** [comment]
Uhmm interesting I'll reopen this and I'll take a look as soon as I can. Thanks for reporting it @pcoltau.

**Comment by Vkt0r** [comment]
@pcoltau The issue has been addressed in #379.
